src/ott/solvers/quadratic/histogram_transport.py

```diff
@@ -0,0 +1,243 @@
+
+ProgressCallbackFn_t = Callable[
+    [Tuple[np.ndarray, np.ndarray, np.ndarray, "HTState"]], None]
+
+
+class HTOutput(NamedTuple):
+  """Holds the output of the Histogram Transport solver.
+
+  Args:
+  costs:
+  linear_convergence:
+  converged:
+  errors:
+  linear_state:
+  geom:
+  old_transport_mass:
+  """
+
+  converged: bool = False
+  errors: Optional[jnp.ndarray] = None
+  linear_state: Optional[LinearOutput] = None
+  geom: Optional[geometry.Geometry] = None
+
+  def set(self, **kwargs: Any) -> "HTOutput":
+    """Return a copy of self, possibly with overwrites."""
+    return self._replace(**kwargs)
+
+  @property
+  def matrix(self) -> jnp.ndarray:
+    """Transport matrix."""
+    return self.linear_state.matrix
+
+  def apply(self, inputs: jnp.ndarray, axis: int = 0) -> jnp.ndarray:
+    """Apply the transport to an array; axis=1 for its transpose."""
+    return self.linear_state.apply(inputs, axis=axis)
+
+  @property
+  def reg_ht_cost(self) -> float:
+    """Regularized optimal transport cost of the histogram transport."""
+    return self.linear_state.reg_ot_cost
+
+  @property
+  def primal_cost(self) -> float:
+    """Return transport cost of current linear OT solution at geometry."""
+    return self.linear_state.transport_cost_at_geom(other_geom=self.geom)
+
+
+class HTState(NamedTuple):
+  """State of the Histogram-Transport solver.
+
+  Attributes:
+  costs:
+  linear_convergence:
+  linear_state:
+  linear_pb:
+  rngs:
+  errors:
+  """
+
+  costs: jnp.ndarray
+  linear_convergence: jnp.ndarray
+  linear_state: LinearOutput
+  linear_pb: linear_problem.LinearProblem
+  rngs: Optional[jax.random.PRNGKeyArray] = None
+  errors: Optional[jnp.ndarray] = None
+
+  def set(self, **kwargs: Any) -> "HTState":
+    """Return a copy of self, possibly with overwrites."""
+    return self._replace(**kwargs)
+
+  def update(  # noqa: D102
+      self,
+      iteration: int,
+      linear_sol: LinearOutput,
+      linear_pb: linear_problem.LinearProblem,
+      store_errors: bool,
+  ) -> "HTState":
+    costs = self.costs.at[iteration].set(linear_sol.reg_ot_cost)
+    errors = None
+    if store_errors and self.errors is not None:
+      errors = self.errors.at[iteration, :].set(linear_sol.errors)
+    linear_convergence = self.linear_convergence.at[iteration].set(
+        linear_sol.converged
+    )
+
+    return self.set(
+        linear_state=linear_sol,
+        linear_pb=linear_pb,
+        costs=costs,
+        linear_convergence=linear_convergence,
+        errors=errors,
+    )
+
+
+# @jax.tree_util.register_pytree_node_class
+class HistogramTransport:
+  """Histogram Transport solver.
+
+  Computes the First Lower Bound distance between two distributions.
+  The current implementation requires uniform marginals.
+  If there are an uneven number of points in the two distributions,
+  then we perform a stratified subsample of the distribution of
```

Comment:
i am not sure a stratified sample is the way to go here. I think something a lot simpler could be to use (soft) quantiles, and pass the quantile levels with a default of, e.g., `jnp.linspace(0,1,num=100)` for instance. 

also, I think we're not taking into account point cloud weights here? (i.e. `a`, `b`). I think it's ok to have a solution that only works for uniform, but this must be more explicit.

Reply:
- So I actually initially tried with `jax.lax.quantile`, but found that it significantly increased the time (using the default interpolation method).
- We are not taking into account point cloud weights here, which I find pretty sad! I do have an implementation for variable weights based on the implementation in `scipy` [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html), but it drops off in speed relative to entropic GW at around 500 data points, so I didn't put it in this PR.
- But yeah, I will try to make the documentation here more explicit

Reply:
Update on this: In my most recent implementation with quantiles, it actually went at the same speed. I have no clue what was wrong with my other implementation.